weblab/fitting/views.py

```diff
@@ -197,3 +204,150 @@ def get(self, request, *args, **kwargs):
+
+        if is_new:
+            if queued:
+                messages.info(self.request, "Fitting experiment submitted to the queue.")
+            else:
+                messages.error(self.request, "Fitting experiment could not be run: " + self.runnable.return_text)
+        else:
+            messages.info(self.request, "Fitting experiment was already run.")
+
+        return super().form_valid(form)
+
+    def get_success_url(self):
+        if hasattr(self, 'runnable'):
+            return reverse('fitting:result:version', args=[self.runnable.fittingresult.pk, self.runnable.pk])
+
+
+class FittingResultFilterJsonView(LoginRequiredMixin, PermissionRequiredMixin, View):
+    """
+    JSON view of valid fitting result input values based on those already selected
+
+    For example, if a model id is specified (as a GET param), only versions of that model
+    (which are visible to the user) will be included in the results. Otherwise all visible
+    models and versions will be in the results (which are simply a list of database
+    IDs of the relevant objects)
+
+    Connections between protocols, fitting specs and datasets are also enforced.
+    """
+    permission_required = 'fitting.run_fits'
+
+    def get(self, request, *args, **kwargs):
+        options = {}
+
+        models = ModelEntity.objects.visible_to_user(request.user)
+        model_versions = CachedModelVersion.objects.visible_to_user(request.user)
+        protocols = ProtocolEntity.objects.visible_to_user(request.user)
+        protocol_versions = CachedProtocolVersion.objects.visible_to_user(request.user)
+        fittingspecs = FittingSpec.objects.visible_to_user(request.user)
+        fittingspec_versions = CachedFittingSpecVersion.objects.visible_to_user(request.user)
+        datasets = Dataset.objects.visible_to_user(request.user)
+
+        def _get_int_param(fieldname, _model):
+            try:
+                pk = int(request.GET.get(fieldname, ''))
+                return _model.objects.get(pk=pk)
+            except ValueError:
+                pass
+            except ObjectDoesNotExist:
+                pass
+
+        model = _get_int_param('model', ModelEntity)
+        protocol = _get_int_param('protocol', ProtocolEntity)
+        fittingspec = _get_int_param('fittingspec', FittingSpec)
+        dataset = _get_int_param('dataset', Dataset)
+
+        if not protocol:
+            if fittingspec:
+                protocol = fittingspec.protocol
+            elif dataset:
+                protocol = dataset.protocol
+
+        # Restrict to specified model and its versions
+        if model:
+            models = [model]
+            model_versions = model_versions.filter(entity__entity=model)
+
+        # Restrict to specified fitting spec and its versions
+        if fittingspec:
+            fittingspecs = [fittingspec]
+            fittingspec_versions = fittingspec_versions.filter(entity__entity=fittingspec.id)
```

Comment:
Is there a reason `.id` is used here and not for model above?

Reply:
No reason, just an oversight! :)